pkg/workflow/dataimpl.go

```diff
@@ -11,62 +14,158 @@ import (
 
 // DataImpl is the default implementation of the Data interface.
 type DataImpl struct {
-	identifier Identifier
-	header     http.Header
-	payload    interface{}
-	errors     []snyk_errors.Error
+	identifier      Identifier
+	header          http.Header
+	contentType     string
+	inputData       Data
+	payload         interface{}
+	payloadLocation Location
+	configuration   configuration.Configuration
+	logger          *zerolog.Logger
+	errors          []snyk_errors.Error
+}
+
+var _ Data = (*DataImpl)(nil)
+
+type PayloadLocation int
+
+const (
+	InMemory PayloadLocation = iota
+	OnDisk
+)
+
+type Location struct {
+	Path string
+	Type PayloadLocation
 }
 
 const (
 	Content_type_key     string = "Content-Type"
 	Content_location_key string = "Content-Location"
 )
 
-// NewDataFromInput creates a new data instance from the given input data.
-//
-// It will preserve the headers, metadata and errors
-func NewDataFromInput(input Data, typeIdentifier Identifier, contentType string, payload interface{}) Data {
-	if len(typeIdentifier.Path) <= 0 {
-		panic("Given identifier is not a type identifier")
+type Option = func(d *DataImpl)
+
+func withIdentifier(id *Identifier) Option {
+	return func(d *DataImpl) {
+		d.identifier = *id
+	}
+}
+
+func withContentType(contentType *string) Option {
+	return func(d *DataImpl) {
+		d.contentType = *contentType
+	}
+}
+
+func withPayload(payload *interface{}) Option {
+	return func(d *DataImpl) {
+		d.payload = *payload
+	}
+}
+
+func WithConfiguration(config *configuration.Configuration) Option {
+	return func(d *DataImpl) {
+		d.configuration = *config
```

Comment:
Suggestion: For all `WithXyz(xyz)` you don't have to hold on to everything that is given to the with function, you can also derive the values in with and not hold onto the configuration, the data ... This way you enable to release memory earlier and don't make a data object container for everything. In some reasons it might be useful but not as a default.